### ISSUE TITLE
Support drop time series if contains more than maxLabelsPerTimeseries

### ISSUE THREAD

**Issue by bertinm-gc** [VictoriaMetrics/VictoriaMetrics#6928]
### Is your feature request related to a problem? Please describe

When sending a metric with over 30 labels (or a custom limit) victoria-metrics drops the labels but the time series remains.
I would like to have the option to configure vmagent to drop the series instead of victoria-metrics dropping the labels, and prevent them from reaching victoria-metrics entirely.

### Describe the solution you'd like

I think we need to add 2 flags to vmagent:

1. The first should be whether to drop the time series when reaching max label `dropTimeseriesMaxLabels` and is a boolean.
2. The second should be an int with the limit of max labels with a default of 30, like victoria-metrics default max labels. `maxLabelsPerTimeseries `

### Describe alternatives you've considered

I considered a solution of setting a flag on victoria-metrics to drop timeseries instead of labels, but I think it would be best to drop the timeseries before it reaches victoria-metrics to reduce the load on victoria-metrics.

I also tried to implement this using relabel config on vmagent, but was unable to find a way to calculate at this level how many labels exists and to drop based on that.

**Comment by Haleygo** [comment]
I agree it should be implemented on agent side.
There is a similar limit in [prometheus](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config) when scraping:
```
# Per-scrape limit on number of labels that will be accepted for a sample. If
# more than this number of labels are present post metric-relabeling, the
# entire scrape will be treated as failed. 0 means no limit.
[ label_limit: [<int>](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#int) | default = 0 ]
```
It will drop all metrics exposed by the target if any metric exceeds the label limit.
Compare to only dropping problematic metrics, this approach makes debugging easier, as users will notice the issue more quickly and identify the problematic metrics in scrape error messages. Does that work for your case?

**Comment by Haleygo** [comment]
After disscusion with the team, we decided to add the `label_limit` option. Both at the global level as `-promscrape.labelLimitPerTarget` flag, and `sample_limit` field in the [scrape config](https://docs.victoriametrics.com/sd_configs/#scrape_configs).
vmagent should drop the metric if its labels exceeded the `label_limit`, report it with a new metric `vm_promscrape_max_label_limit_exceeded_errors_total` and log messages to inform the user.

**Comment by hartfordfive** [comment]
So if I understand this correctly, if a scrape exceeds the configured `samples_limit` or `series_limit` on the scrape job config, the entire scrape will fail, although this same functionality isn't yet available for `labels_limit`?

**Comment by hagen1778** [comment]
>vmagent should drop the metric if its labels exceeded the label_limit, report it with a new metric vm_promscrape_max_label_limit_exceeded_errors_total and log messages to inform the user.

@Haleygo Prometheus docs suggest that the whole scrape should fail in this case. Did you discuss this behavior and why it was decided to drop the individual sample instead?

**Comment by Haleygo** [comment]
> > vmagent should drop the metric if its labels exceeded the label_limit, report it with a new metric vm_promscrape_max_label_limit_exceeded_errors_total and log messages to inform the user.
> 
> @Haleygo Prometheus docs suggest that the whole scrape should fail in this case. Did you discuss this behavior and why it was decided to drop the individual sample instead?

@hagen1778 We don't see the necessity of dropping the entire target when only few metrics need to be fixed, and the rest remain valid.
Also, vmagent uses [stream parsing mode](https://docs.victoriametrics.com/vmagent/#stream-parsing-mode) for scraping big target, and metrics from single target are pushed to remote storage in multiple requests. So the whole scrape can't be discarded if one metric exceeds the label limit.

@hartfordfive An alerting rule for `vm_promscrape_max_label_limit_exceeded_errors_total` can help identify targets  sending metrics that exceed the limit, and log in vmagent should expose the problematic metric. Or do you have any other concerns about using the`soft` mode?

**Comment by hartfordfive** [comment]
To be fair, I agree the default label limit of 30 per time series should be more than sufficient.   My concern is that if someone chooses to lower that limit, the _soft_ implementation of limiting labels may not be as obvious to identify problematic targets and this could lead to skewing of some statistics on some dashboards until the issue is resolved.  

For example, a scenario where a fictional service reads logs from a Kafka topic to then store the data in a given storage location.   The relevant counts would be exposed via the `log_archiver_output_events_total` counter metric with the following alphabetically sorted labels:
- app
- category
- dest_location
- event_type
- src_topic

An administrator could choose to limit the number of labels to 5 :
```
log_archiver_output_events_total{app="website", category="frontend", event_type="persist", src_topic="archive-logs-website", dest_location="disk-file"}
log_archiver_output_events_total{app="website", category="frontend", event_type="persist", src_topic="archive-logs-website", dest_location="disk-file"}

log_archiver_output_events_total{app="postgres-audit", category="db", event_type="persist", src_topic="archive-logs-postgres-audit", dest_location="disk-file"}
log_archiver_output_events_total{app="postgres-audit", category="db", event_type="persist", src_topic="archive-logs-postgres-audit", dest_location="disk-file}
```

Now if at a certain point later on someone decides to update the metrics to add a new `status` label which can have the values either `success` or `fail`.
```
log_archiver_output_events_total{app="website", category="frontend", event_type="persist", src_topic="archive-logs-website", dest_location="disk-file", status="success"}
log_archiver_output_events_total{app="website", category="frontend", event_type="persist", src_topic="archive-logs-website", dest_location="disk-file", status="error"}
log_archiver_output_events_total{app="website", category="frontend", event_type="discard", src_topic="archive-logs-website", dest_location="disk-file", status="success"}

log_archiver_output_events_total{app="postgres-audit", category="db", event_type="persist", src_topic="archive-logs-postgres-audit", dest_location="disk-file", status="success"}
log_archiver_output_events_total{app="postgres-audit", category="db", event_type="persist", src_topic="archive-logs-postgres-audit", dest_location="disk-file", status="fail"}
log_archiver_output_events_total{app="postgres-audit", category="db", event_type="discard", src_topic="archive-logs-postgres-audit", dest_location="disk-file", status="success"}
```

This extra `status` label gives an important breakdown of the number of events have successfully been persisted vs those which haven't.   So in a case where vmagent is configured to sort the labels alphabetically, given a labels limit of 5, the new `status` label gets dropped and thus becomes impossible to differentiate between successful and failed output events.   Should an alerting condition such as this one get pushed ahead of time in anticipation of the new label:
```
increase(log_archiver_output_events_total{status="fail"}[10m]) >= 10
```
It would never fire even once the updated metrics are exposed as the 6th label `status` would be dropped.  A Grafana panel graphing a query such as the one below may not immediately give the impression of a problem as it would still load samples:
```
sum(
  rate(log_archiver_output_events_total[5m])
) by (app, event_type, status) 
```
Dropping all metrics from the target becomes much more obvious obvious and avoid possibilities where random labels are stored if it's decided not to sort them alphabetically.

**Comment by hagen1778** [comment]
> To be fair, I agree the default label limit of 30 per time series should be more than sufficient. My concern is that if someone chooses to lower that limit, the soft implementation of limiting labels may not be as obvious to identify problematic targets and this could lead to skewing of some statistics on some dashboards until the issue is resolved.

I agree with your point. But there are two ways how you can prevent this from happening:
1. You can get notified about [labels being dropped on ingestion](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/a72e1155b90018369e8c79e872e130b429ccc482/deployment/docker/alerts-cluster.yml#L143-L153).
2. You can check whether your alerting rule match something (and alert if not) via https://docs.victoriametrics.com/vmalert/#never-firing-alerts

**Comment by hagen1778** [comment]
> Also, vmagent uses [stream parsing mode](https://docs.victoriametrics.com/vmagent/#stream-parsing-mode) for scraping big target, and metrics from single target are pushed to remote storage in multiple requests. So the whole scrape can't be discarded if one metric exceeds the label limit.

What about metrics pushed into vmagent? Should we have a separate cmd-line flag for them?

**Comment by hagen1778** [comment]
Looks like the most sensible way would be to change behaviour of `-maxLabelsPerTimeseries` cmd-line flags to drop the whole time series, instead of dropping labels. Dropping labels could result into series collision and bring additional harm on query path. If someone has arguments regarding usefulness of dropping labels instead of series - please add your comment here.

The `-maxLabelsPerTimeseries` should also be available on the vmagent side, so bad data could be dropped as soon as possible.

I aslo agree with this comment https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6928#issuecomment-2328049461 - the similar option to drop series with many labels should be available on promscrape level. This setting couldn't exceed the `-maxLabelsPerTimeseries` though. This setting won't drop the whole scrape, as Prometheus does, because vmagent can parse scrape response in stream mode where scrape can't be rejected as a whole.

**Comment by timurbabs** [comment]
How about keep a way with dropping labels and add an option to specify the priority of labels, allowing users to define which labels should be dropped first if the label count exceeds the configured limit? This feature would help ensure that essential labels are retained when limits are reached, providing greater control over data management.
Also we can keep the option with dropping the whole time series in a separate flag.

**Comment by hagen1778** [comment]
@timurbabs this sounds too complicated for users to maintain, especially on the global level, so I don't think anyone will use it.
The same behavior can be achieved via https://docs.victoriametrics.com/relabeling/, but I don't see many people doing it.

**Comment by hagen1778** [comment]
Related tickets https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6817

**Comment by f41gh7** [comment]
The feature was included into [v1.108.0](https://docs.victoriametrics.com/changelog/#v11080) release.
